### DISCUSSION TITLE
How to read from storage buffer in vertex shader

### DISCUSSION THREAD

**Discussion by AshleyScirra** [gpuweb/gpuweb#1402] [Q&A] [answered]
I don't have much experience using storage buffers specifically previously, but I was trying to experiment with vertex pulling, i.e. having the shader read vertices from buffer indices it calculates itself.

I think I got the bindings set up OK, but when I try to read from a storage buffer in WGSL, I get a tint error in the current Chrome Canary. WGSL code:

```
[[block]] struct BufferData {
    pos : vec4<f32>;
};
[[binding(0), set(0)]] var<storage> bufferData : [[access(read)]] BufferData;

[[builtin(position)]] var<out> Position : vec4<f32>;
[[builtin(vertex_index)]] var<in> VertexIndex : u32;

[[stage(vertex)]]
fn main() -> void {
    Position = bufferData[VertexIndex].pos;
    return;
}`;
```

Canary logs in the console:
```
Tint WGSL reader failure:
Type Determination: error: 12:26: invalid parent type (__struct_tint_symbol_2) in array accessor
```
I guess it's saying I can't array index in to `bufferData`. Have I got this wrong? I can't find any samples that cover this and it's hard to piece it together from the spec.

**Comment by AshleyScirra** [comment, marked as answer]
Answering my own question but think I figured out it's this:

```
[[block]] struct BufferData {
	[[offset(0)]] pos : [[stride(16)]] array<vec4<f32>>;
};
[[binding(0), set(0)]] var<storage> bufferData : [[access(read)]] BufferData;
```

and then read with `bufferData.pos[VertexIndex]`.

**Reply by Kangz** [reply]
Yeah the WGSL error messages in Chromium are still in progress. Your solution is correct!

**Reply by AshleyScirra** [reply]
Thanks for confirming!
